futures-async-macro/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+            })
+            .to_tokens(tokens);
+            expr.to_tokens(tokens);
+        });
+    });
+
+    tokens.into()
+}
+
+/// The scope in which `#[for_await]`, `await!` and `await_item!` was called.
+///
+/// The type of generator depends on which scope is called.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Scope {
+    /// `async fn`, `async {}`, `async ||`
+    Future,
+    /// `#[async_stream]`, `async_stream_block!`
+    Stream,
+    /// `static move ||`, `||`
+    ///
+    /// It cannot call `#[for_await]`, `await!` and `await_item!` in this scope.
+    Closure,
+}
+
+use Scope::{Closure, Future, Stream};
+
+struct Expand(Scope);
+
+impl Expand {
+    /// Expands `#[for_await] for pat in expr { .. }`.
+    fn expand_for_await(&self, mut expr: ExprForLoop) -> Expr {
+        if !(expr.attrs.len() == 1 && expr.attrs[0].path.is_ident("for_await")) {
+            return Expr::ForLoop(expr);
+        } else if !expr.attrs[0].tts.is_empty() {
+            return error::expr_compile_error(&Error::new_spanned(
+                expr.attrs.pop(),
+                args_is_not_empty!("for_await"),
+            ));
+        }
+
+        // It needs to adjust the type yielded by the macro because generators used internally by
+        // async fn yield `()` type, but generators used internally by `async_stream` yield
+        // `Poll<U>` type.
+        let yield_ = match self.0 {
+            Future => TokenStream2::new(),
+            Stream => quote! { ::futures::core_reexport::task::Poll::Pending },
+            Closure => return outside_of_async_error!(expr, "#[for_await]"),
+        };
+        let ExprForLoop { label, pat, expr, body, .. } = expr;
+
+        // Basically just expand to a `poll` loop
+        syn::parse_quote! {{
+            let mut __pinned = #expr;
+            #label
+            loop {
+                let #pat = {
+                    match ::futures::async_stream::poll_next_with_tls_context(unsafe {
+                            ::futures::core_reexport::pin::Pin::new_unchecked(&mut __pinned)
+                        })
+                    {
+                        ::futures::core_reexport::task::Poll::Ready(e) => {
+                            match e {
+                                ::futures::core_reexport::option::Option::Some(e) => e,
+                                ::futures::core_reexport::option::Option::None => break,
+                            }
+                        }
+                        ::futures::core_reexport::task::Poll::Pending => {
+                            yield #yield_;
+                            continue
+                        }
+                    }
+                };
+
+                #body
+            }
+        }}
+    }
+
+    /* TODO: If this is enabled, it can be used for `yield` instead of `stream_yield!`.
+             Raw `yield` is simpler, but it may be preferable to distinguish it from `yield` called
+             in other scopes.
```

Comment:
The main downside of using `yield` expressions instead of a `stream_yield!` macro is that the proc-macro can't see through any other macros called in the passed `TokenStream`, so it becomes impossible for users to define their own macros that expand to `yield` expressions intended to be used as part of an `async_stream`.

That said, when I implemented `embrio_async::async_stream_block` I went for just matching `yield` expressions, because it makes it much cleaner and it's unlikely that users will really need to use macros expanding to `yield`.

Reply:
Actually, it looks like `stream_yield!` is just a token being matched by the `async_stream` macro, not a `macro_rules!`-macro? Then you don't get the upside of allowing users to define macros that expand to `stream_yield!` and just the downside of being weird and harder to read, so I'd definitely suggest changing to just expand `yield` expressions.

Reply:
Oh, it's both, so yeah, not sure which downside is the bigger one.

Reply:
I changed it to use the `yield` expression. Because we can analyze what is in the scope (but it takes some effort) of the [attr](https://github.com/taiki-e/auto_enums/blob/3c88acc88ecb1debf8f02976fa1fcf2ef3c592a4/core/src/auto_enum/attrs.rs#L9)/[macro](https://github.com/taiki-e/futures-rs/blob/aca5bc00cfd26b0d5a786eeeee80077a10db54ab/futures-async-macro/src/lib.rs#L297), so I think it is better to use something that is easy for the user to use.